agent/util-scripts/pbench-tool-meister-start

```diff
@@ -1215,22 +1110,44 @@ if __name__ == "__main__":
         default=None,
         help="The list of system information items to be collected.",
     )
+    parser.add_argument(
+        "--orchestrate",
+        dest="orchestrate",
+        default="create",
+        choices=("create", "existing"),
+        help=(
+            "The `create` keyword directs the command to create the various"
+            " instances of the Redis server, Tool Data Sink, and Tool"
+            " Meisters, while the `existing` keyword directs the command to"
+            " use existing instances of all three."
+        ),
+    )
```

Comment:
I still think that this would be better as a binary (True/False) option rather than a keyword "choice".  If we call it something like `--external-orchestration` and give it a default value of false, then, unless the switch is specified (that is, for all cases where we're not doing external orchestration), pbench will manage the processes.

If we're really in love with having a keyword interface for the UI (i.e., because we honestly think that we might expand beyond two choices), then that's fine, but, in this proposed implementation, we shouldn't be propagating the keyword-ness beyond the command line processing:  after that point, we should have the state captured in a boolean.  Having tests like `if orchestrate == "create"` is a bad idea, because it's too easy to misspell the keyword (or to be inconsistent about which of the choices we reference); and, having different instances of the test in various parts of the code is just asking for problems.

Reply:
Philosophically, I'd like to avoid using boolean for command line options and use enums instead.

I can make an enum of the choices and use that for the tests.

Reply:
Yeah...at some point, I would love to have you explain why booleans are to be avoided!  To me they are easy to understand and hard to screw up...at least when the choice is truly binary.

Anyway, if you really want to avoid a "switch", then just encapsulate all the `orchestrate == "create"` tests in a function which returns a boolean value, and I'll shut up.

Reply:
> Yeah...at some point, I would love to have you explain why booleans are to be avoided!

See https://www.teamten.com/lawrence/programming/prefer-enums-over-booleans.html, https://betterprogramming.pub/dont-use-boolean-arguments-use-enums-c7cd7ab1876a, https://nedbatchelder.com/blog/200507/booleans_suck.html regarding.

> just encapsulate all the orchestrate == "create" tests in a function which returns a boolean value

Good idea.

Reply:
Thanks for the references -- those were interesting reading.

Unfortunately, the last one torpedoes his own argument at the end where he raises my exact concerns over _not_ using a boolean.  The second one was also only marginally convincing:  clearly, if you are representing a series of "states", an enum is a better choice; however, that approach suffers when the states are not properly sequenced or distinct, and it is still subject to the evolution problem outlined at the end of the last reference, where you have to reevaluate all of your existing logic when you add the third (or even other) states.  I found the first reference to be the most useful, but, in actuality, his principal argument is just a strawman (I sure that it was a real mistake, but the mistake was not in using a boolean but rather in the understanding of what the boolean meant).

So, it seems like that concerns with booleans are:

1. Lack of extensibility
2. Lack of clarity as to what the two values mean
3. Lack of readability of function invocations
4. Lack of type safety

Lack of extensibility is not necessarily a "problem" -- in fact, it can be a "feature".  It is a problem when the interface is poorly designed or conceptualized; if the concept being represented truly is a binary choice, then using a boolean _supports_ the design and _avoids_ problems in the code.  (This is my principal reason for recommending them, here.)

Lack of clarity also reflects a problem in the design or conceptualization, not in the datatype choice.  Enums can come to the rescue here by renaming "True" and "False" to something interface-specific, but this issue can also be addressed in other ways (such as by naming the interface elements properly or by providing explicit testing functions rather than by testing the attribute directly).  (So, I don't see this alone as a reason to give up the rigor of boolean operation.)

Lack of readability depends on the language and coding style.  In a language like Python which supports keyword arguments, by defining good parameter names and by using them in function invocations, the intention of the caller is clear.  Obviously, using enums would do the name thing, but at the cost of additional mechanism.  (So, for us, working in Python, this is not a reason to give up booleans.)

Lack of type safety would be a good argument, but, if we assume that we're using keyword arguments, there's no possibility of mixing up the parameters.  (So, enums buys us nothing here.)

Therefore, to me, it comes down to trade-off between extensibility and rigorous implementation.  If what we're describing or implementing really has only two states, either has or doesn't have a particular characteristic, or is otherwise represented by a binary choice, then a boolean is unquestionably the right approach -- it just cannot be screwed up, providing we use good engineering (without which enums won't save us...).  On the other hand, if what we're describing or implementing could reasonably be expected to evolve a third state or is otherwise a little squishy, then an enum would be a better choice.

Either way, "rolling our own enums" (e.g., using strings) is a bad idea:  it has type un-safety built right in, and it suffers from all of the growth problems.  We can make that approach more robust, by encapsulating it behind a functional interface for setting and testing the values, but, really, if that's where we're going, we should just switch to an enum.  (Perhaps in the next PR....)

Reply:
I think the big concern here is that the evolution from strict boolean (yes we orchestrate, no we don't) is not an enum, but rather a set of enums (or else a proliferation of confusing enums). That is, we want to orchestrate Redis, or the data sync, or the tool meisters; or data sync and tool meisters, or Redis and tool sync but not tool meisters (a plausible case if tool meisters are fired up on cluster pods but the external controller wants us to take care of the data sync they'll use).... etc.

I'm not sure we've really thought out all of the use cases and tradeoffs... which plays back into why Peter didn't want to take on more than the boolean segregation now.

Reply:
Understood, and I concur...and, I think you just explained why `--orchestrate` should _not_ take a keyword/choice:  in a conceivable future, the question of whether or not Pbench should create the Redis server will be orthogonal to the question of whether or not it should create the data sink or the tool meisters, so trying to represent all those (overlapping) states with a single enum/option will likely be problematic, while representing them with a series of booleans will be clear.  (Yes, using booleans _would_ result in the possibility of invalid combinations, but these are easily detected during command line parsing; this is balanced against the challenge of enumerating all of the possible valid combinations and documenting the "name" of each of them.)

So, I don't think it's wrong for `--orchestrate` (or, rather, something like `--external-orchestration`) to be a boolean:  at the moment, it would mean "all or nothing"; in the future, we could split off other components to their own switches, or add semantics to the existing switches which describe the other components.